pkg/virt-operator/resource/apply/rbac.go

```diff
@@ -4,183 +4,250 @@ import (
 	"context"
 	"fmt"
 
+	"github.com/openshift/library-go/pkg/operator/resource/resourcemerge"
+	"k8s.io/apimachinery/pkg/api/equality"
+
+	"k8s.io/client-go/tools/cache"
+
+	"kubevirt.io/kubevirt/pkg/controller"
+
 	rbacv1 "k8s.io/api/rbac/v1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 
 	"kubevirt.io/client-go/log"
 	"kubevirt.io/kubevirt/pkg/virt-operator/resource/generate/rbac"
 )
 
+type RoleType int
+
+const (
+	TypeRole               RoleType = iota
+	TypeClusterRole        RoleType = iota
+	TypeRoleBinding        RoleType = iota
+	TypeClusterRoleBinding RoleType = iota
+)
+
 func (r *Reconciler) createOrUpdateClusterRole(cr *rbacv1.ClusterRole, imageTag string, imageRegistry string, id string) error {
+	return r.createOrUpdate(cr, imageTag, imageRegistry, id, TypeClusterRole, false)
```

Comment:
I think this should still be 4 separate functions or at a minimum 2. If you look at other objects that are reconciled they are all quite similar. (ex. deployment and daemonset) but for simplicity and readability we have kept them separate. This allows us to easily know exactly what is happening with each resource as well as if in the future we need to change the functionality of one and not the other that is easy as well. If they need the same thing its easy to create a function and call it in both. The `createOrUpdate` function is very hard to ready now and requires many manual stops and checks to go check functions. such as `getRoleCreateFunction` which can return a create function for a role, but also a clusterrole and also bindings of both type which are four separate resources. `areRoleRulesEqual` function is also hard to ready from a parent level and understand, if we are comparing a binding what rules are we checking? does this also check cluster rules? 

Also i think we can remove passing in `imageTag`, `imageRegistry` and `id` to these function and instead get it in the function from the `r.kv` object or am I missing something? I think it is passing in values returned from `getTargetVersionRegistryID` ?

Reply:
I totally get your point but I'm not sure I that fully agree. My thoughts:
- Most of the private helper functions (like `getRoleTypeName`, `getRoleMetaObject`, `getRoleCreateFunction` etc') are totally straight forward and obvious. For more clarity I can remove the "role" part from the name and call them `getName`, `getMetaObject`, `getCreateFunction`. It's true that these functions operate on completely different resources, but because we're talking about private helper functions within a certain context, I don't think that it's problematic. After all, rbacbackup deals with those four resources, so they are related within the local context of rbacbackup IMHO.
- `areRoleRulesEqual` is the only non-trivial one because there's different behavior for roles and for binding. But still I think that the function itself is very short, simple and easy to understand.
- Having 2 functions is problematic here since we would still need to have all of the helper functions.

As I see it the question we need to ask is what's more probable (in the context of rbacbackup):
1) That we need to change the behavior in the same way for all resources (which happened in this reconciliation task)
2) That we need to change the behavior of only one of the resources or change every resource's behavior in a different way.

It seems to me that the first one is more probable and therefore I'm not for having 4 functions that are 99% duplicated. Also I think that if we do get to a point that the implementation for every resource is very different it's very easy to copy-paste and duplicate the code again. In the mean time though, I don't think that every change should appear in 4 different places.

Does that make sense to you?

Reply:
Regarding the passing of `imageTag`, `imageRegistry` and `id` - totally agree :) thanks